### PR TITLE
 Updated Read operation with TTL and error handling on Android

### PR DESCRIPTION
**Pull request by jackycodes**
Things to consider before you submit the PR:

* [ ] Has `CHANGELOG.md` been updated?
* [x] Are tests passing locally?
* [x] Are the files formatted correctly?
* [ ] Did you test your change with either the sample apps that are included in the repository or with a blank app that uses your change?

## Description

1. Updated Read operation with:
  - option to pass a time to live
  - error handling
2. Moved utils to its own class

### CODE REVIEW

File: appcenter-data/Data.js [hunk 1]
```diff
@@ -5,10 +5,26 @@ const ReactNative = require('react-native');
...
+    APP_DOCUMENTS = 'readonly'
+};
+
 const Data = {
-    read(documentId, partition) {
-        return AppCenterReactNativeData.read(documentId, partition).then(result => {
+    TimeToLive,
+    DefaultPartitions,
+    read(documentId, partition, ttl) {
+        if(ttl === undefined) {
```
Comment (guperrot) [inline appcenter-data/Data.js]: this if does not look its formatted correctly
```suggestion
        if (ttl === undefined) {
```
I am curious why eslint didnt see it.

File: appcenter-data/android/src/main/java/com/microsoft/appcenter/reactnative/data/AppCenterReactNativeDataUtils.java [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+import com.facebook.react.bridge.WritableNativeArray;
+import com.facebook.react.bridge.WritableNativeMap;
+
+import com.google.gson.JsonArray;
+import com.google.gson.JsonElement;
+import com.google.gson.JsonObject;
+import com.google.gson.JsonPrimitive;
+
+public class AppCenterReactNativeDataUtils {
+    public static WritableMap convertJsonObjectToWritableMap(JsonObject jsonObject) {
```
Comment (guperrot) [inline appcenter-data/android/src/main/java/com/microsoft/appcenter/reactnative/data/AppCenterReactNativeDataUtils.java]:
```suggestion

    public static WritableMap convertJsonObjectToWritableMap(JsonObject jsonObject) {
```

File: appcenter-data/android/src/main/java/com/microsoft/appcenter/reactnative/data/AppCenterReactNativeDataModule.java [hunk 3]
```diff
@@ -55,11 +56,16 @@ public String getName() {
...
-    public void read(String documentId, String partition, final Promise promise) {
-        Data.read(documentId, JsonElement.class, partition).thenAccept(new AppCenterConsumer<DocumentWrapper<JsonElement>>() {
+    public void read(String documentId, String partition, int ttl, final Promise promise) {
+        Data.read(documentId, JsonElement.class, partition, new ReadOptions(ttl)).thenAccept(new AppCenterConsumer<DocumentWrapper<JsonElement>>() {
 
             @Override
             public void accept(DocumentWrapper<JsonElement> documentWrapper) {
+                if (documentWrapper.getError() != null) {
+                    DataException dataException = documentWrapper.getError();
+                    promise.reject("Read failed", dataException);
```
Comment (guperrot) [inline appcenter-data/android/src/main/java/com/microsoft/appcenter/reactnative/data/AppCenterReactNativeDataModule.java]: Did you test how the exception looks like in Javascript? What properties it contains?

File: appcenter-data/android/src/main/java/com/microsoft/appcenter/reactnative/data/AppCenterReactNativeDataUtils.java (hunk 2 above)
Comment (jackycodes) [reply appcenter-data/android/src/main/java/com/microsoft/appcenter/reactnative/data/AppCenterReactNativeDataUtils.java]: Fixed

File: appcenter-data/Data.js (hunk 1 above)
Comment (jackycodes) [reply appcenter-data/Data.js]: For some reason working at the root of the project from VSCode doesn't pick it up and eslint errors. If I set my workspace to a specific project eslint errors will show up. I'll be sure to run the eslint script from now on.

File: appcenter-data/android/src/main/java/com/microsoft/appcenter/reactnative/data/AppCenterReactNativeDataModule.java (hunk 3 above)
Comment (jackycodes) [reply appcenter-data/android/src/main/java/com/microsoft/appcenter/reactnative/data/AppCenterReactNativeDataModule.java]: Something like this for an invalid documentId:

Screen Shot 2019-06-21 at 3 21 11 PM

File: appcenter-data/Data.js [hunk 4]
```diff
@@ -5,10 +5,26 @@ const ReactNative = require('react-native');
...
+    USER_DOCUMENTS: 'user',
+    APP_DOCUMENTS: 'readonly'
+};
+
 const Data = {
-    read(documentId, partition) {
-        return AppCenterReactNativeData.read(documentId, partition).then(result => {
+    TimeToLive,
+    DefaultPartitions,
+    read(documentId, partition, ttl) {
```
Comment (guperrot) [inline appcenter-data/Data.js]: Actually we need ReadOptions as an object like on other SDKs.

Comment (dhei) [inline appcenter-data/Data.js]: As discussed on Slack, `timeToLive` should be wrapped inside a `ReadOptions` object rather than passed directly.

Comment (jackycodes) [reply appcenter-data/Data.js]: Added

Comment (jackycodes) [reply appcenter-data/Data.js]: Added

File: appcenter-data/android/src/main/java/com/microsoft/appcenter/reactnative/data/AppCenterReactNativeDataModule.java [hunk 5]
```diff
@@ -56,8 +60,12 @@ public String getName() {
     }
 
     @ReactMethod
-    public void read(String documentId, String partition, int ttl, final Promise promise) {
-        Data.read(documentId, JsonElement.class, partition, new ReadOptions(ttl)).thenAccept(new AppCenterConsumer<DocumentWrapper<JsonElement>>() {
+    public void read(String documentId, String partition, ReadableMap readOptionsMap, final Promise promise) {
+        ReadOptions readOptions = new ReadOptions(TimeToLive.DEFAULT);
```
Comment (guperrot) [inline appcenter-data/android/src/main/java/com/microsoft/appcenter/reactnative/data/AppCenterReactNativeDataModule.java]: avoid allocate an object if being replaced, move it to else.

Comment (jackycodes) [reply appcenter-data/android/src/main/java/com/microsoft/appcenter/reactnative/data/AppCenterReactNativeDataModule.java]: Fixed

**Pull request merged**